NGLess/StandardModules/Parallel.hs

```diff
@@ -416,21 +418,85 @@ pasteHiddenFunction = Function
     , funcAllowsAutoComprehension = False
     }
 
-addHash :: [(Int, Expression)] -> NGLessIO [(Int, Expression)]
-addHash script = do
+addLockHash :: [(Int, Expression)] -> NGLessIO [(Int, Expression)]
+addLockHash script = do
         isSubsample <- nConfSubsample <$> nglConfiguration
-        pureTransform (addHash' isSubsample) script
+        pureTransform (addLockHash' isSubsample) script
     where
-        addHash' :: Bool -> Expression -> Expression
-        addHash' isSubsample (FunctionCall (FuncName fn) expr kwargs block)
-            | fn `elem` ["lock1", "collect"] = FunctionCall (FuncName fn) expr ((Variable "__hash", ConstStr h):kwargs) block
+        addLockHash' :: Bool -> Expression -> Expression
+        addLockHash' isSubsample (FunctionCall (FuncName "lock1") expr kwargs block) =
+            FunctionCall (FuncName "lock1") expr ((Variable "__hash", ConstStr h):kwargs) block
             where
-                h = T.pack . (++ (if isSubsample then "-subsample" else "")) . MD5.md5s . MD5.Str . show $
-                        case fn of
-                            "lock1" -> map snd script
-                            "collect" -> expr:map snd script
-                            _ -> error "impossible case"
-        addHash' _ e = e
+                h = T.pack . (++ (if isSubsample then "-subsample" else "")) . MD5.md5s . MD5.Str . show $ map snd script
+        addLockHash' _ e = e
+
+
+{-| Calculation of hashes for collect method calls
+ so that the hash depends only on the relevant (influencing the
+ collected result) part of the script.
+
+ Hashes for variables are stored in a map (as a state).
+ For each expression (top to bottom) first the block variables
+ are added to the map (if present), then hashes are calculated
+ and applied (in lookups) recursively.
+ Each collect call receives new variable __hash storing the hash
+ of it's own expression (with hashes already applied inside).
+-}
+
+addCollectHashes :: [(Int, Expression)] -> NGLessIO [(Int, Expression)]
+addCollectHashes expr_lst = do
+    env <- nglEnvironment
+    modules <- loadedModules
+    let nVer = ngleVersion env
+    let modInfos = map modInfo modules
+    return $ addCollectHashes' nVer modInfos (M.insert (Variable "ARGV") (T.pack "ARGV") M.empty) expr_lst
+
+addCollectHashes' :: T.Text -> [ModInfo] -> (M.Map Variable T.Text) -> [(Int, Expression)] -> [(Int, Expression)]
+addCollectHashes' _ _ _ [] = []
+addCollectHashes' nV mods state_ ((lno, expr):rest) = case expr of
+        (FunctionCall (FuncName "collect") ex kw block) ->
+            let ((FunctionCall _ _ kw' _), s') = runState (recursiveTransform hashExpression expr) state' in
+                (lno, (FunctionCall (FuncName "collect") ex ((head kw'):kw) block)):(recCall s' rest)
+        _ -> (lno, expr):(recCall (execState (recursiveTransform hashExpression expr) state') rest)
+
+    where
+        state' = execState (recursiveTransform gatherBlockVars expr) state_
```

Comment:
This is not incorrect, but it seems non-canonical. You are manually threading the state, but that is what `State` does.